### PR TITLE
add deployment

### PR DESCRIPTION
**Pull request by tkellen**
Sharing this WIP early. No review needed yet. This is an attempt to simplify our standard Ansible deployment scripts down to the fewest moving parts possible. The Vagrantfile included here is intended purely as test deployment target, it is not used for development.

Tasks remaining:
- [x] automated database dumps
- [x] automated database restoration
- [x] secrets management (will use ansible-vault)

We're still looking good for a Friday proof of concept.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,15 +13,17 @@
...
   "scripts": {
     "build": "webpack",
+    "build:production": "NODE_ENV=production webpack",
+    "build:development": "NODE_ENV=development webpack",
     "migrate": "careen -c backend/db/migrations/careen",
     "migrate:up": "careen -A -e -c backend/db/migrations/careen",
     "migrate:down": "careen -R -e -c backend/db/migrations/careen",
     "migrate:create": "careen -C -c backend/db/migrations/careen",
     "migrate:reset": "careen -R -a -e -n 10000 -c backend/db/migrations/careen; npm run migrate:up",
     "start": "npm-run-all -p -r front:dev server:dev",
```
Comment (gnarf) [inline package.json]: these would need to be updated as well

File: deploy/ansible/provision.yml [hunk 2]
```diff
@@ -0,0 +1,134 @@
+- hosts: all
+  become: yes
+  become_method: sudo
+  gather_facts: no
+  vars_files:
+    - vars/main.yml
+    - "vars/{{env}}.yml"
+  tasks:
+    # Start managing system packages and updates.
+    - name: ensure python is installed
```
Comment (cowboy) [inline deploy/ansible/provision.yml]: Is there any case where python won't be installed?

**Comment by cowboy** [comment]
@tkellen this makes sense to me. Can you add documentation of dependencies and process to the remaining tasks?

File: deploy/ansible/provision.yml (hunk 2 above)
Comment (tkellen) [inline deploy/ansible/provision.yml]: Ubuntu 16 doesn't ship with python installed, which is a prereq for Ansible. We'll always need it for now!

**Comment by tkellen** [comment]
Yes I definitely can. Planning to wrap this up tomorrow.

**Comment by tkellen** [comment]
Secrets and database management are in. I'm going to make SSL a separate PR. Documentation and final testing tomorrow.

**Comment by tkellen** [comment]
If anyone tries this, there is a secure note in lastpass "Image Annotator Vault Password" with the vault passsword.

**Comment by tkellen** [comment]
@gnarf @kadamwhite @boazsender or @cowboy Can ya'll give this a whirl? I'll be writing docs today. To test locally you do:
```
vagrant up
npm run provision:vagrant
npm run deploy:vagrant -- -e commit=deploy
```
Then open http://10.10.0.100 in your browser and you should be able to successfully commit a series of annotations.

**Comment by tkellen** [comment]
PS: If you're prompted for a vault password, it's in lastpass under the Image Annotator note.

**Comment by tkellen** [comment]
TODO:

- [ ] ensure api starts at boot
- [ ] ensure postgresql starts at boot

**Comment by tkellen** [comment]
Documentation is up for this--assuming we are :+1: on it, can you land @gnarf?

**Comment by tkellen** [comment]
@boazsender Would appreciate your review as well!

File: README.md [hunk 3]
```diff
@@ -21,6 +21,40 @@ This project exposes a REST API with the following endpoints:
...
+### npm run edit-secrets
+This will open the secrets file in your default editor. Secret data is managed
+by Ansible Vault. All secrets are stored in `ansible/vars/secrets.yml`.
+
+If you are just getting started on the project, this is the first command
+you should run. You'll need to add yourself as a user in the secrets file,
+submit a PR, and ask someone who already has access to run this for any
+server you wish to have deployment access to.
+
+### npm run provision:[production|staging|vagrant
```
Comment (gnarf) [inline README.md]: nit: missing the closing `]`

File: README.md [hunk 4]
```diff
@@ -21,6 +21,40 @@ This project exposes a REST API with the following endpoints:
 
 More information available in [api.md](blob/master/api.md).
 
+## Deployment
+In order to deploy the site, ensure you have Ansible 2.2x installed. You will
+be prompted for a "Vault password" when running most of these commands. Look
+in the "Image Annotator" note in LastPass for the password.
```
Comment (tkellen) [inline README.md]: Yup!

Comment (tkellen) [inline README.md]: hows this?

In order to deploy the site, ensure you have Ansible 2.2x installed. You will
be prompted for a "Vault password" when running most of these commands. Any of
committer on this project should be able to help!

If you work at Bocoup, you can find the password in the "Image Annotator" note
on our LastPass account.

Comment (boazsender) [inline README.md]: +1

Comment (boazsender) [inline README.md]: maybe remove the bit about Bocoup's lastpass?

**Pull request merged**